### DISCUSSION TITLE
Callbacks for Subscriptions?

### DISCUSSION THREAD

**Discussion by spmatthews03** [davidepianca98/KMQTT#39] [Q&A]
I can see how to subscribe, but how do I call a callback when a subscription is triggered? Does it go inside the client block with a when statement? or somewhere else?

**Comment by davidepianca98** [comment]
Hello, do you mean the callback for when a PUBLISH message is received? Since a PUBLISH can only be received if you have subscribed to a specific topic or topic filter, I assume that's what you need.

You define the callback when instantiating the client just like this:
```kotlin
val client = MQTTClient(
    MQTTVersion.MQTT5,
    "test.mosquitto.org",
    1883,
    null
) {
    // Callback where it is the PUBLISH packet received. You can check the topic of the packet to understand from which subscription you got it
    println(it.topicName)
    println(it.payload?.toByteArray()?.decodeToString())
}
```
